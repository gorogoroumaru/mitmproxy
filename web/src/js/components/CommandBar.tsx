import React, { useState, Component } from 'react'
import classnames from 'classnames'
import { Key, fetchApi } from '../utils.js'

export default function CommandBar() {
    const [command, setCommand] = useState("")
    const [results, setResults] = useState([])
    const [history, setHistory] = useState([])
    const [currentPos, setCurrentPos] = useState(0)

    const onChange = (e) =>  {
        setCommand(e.target.value)
    }

    const onKeyDown = (e) => {
        if (e.keyCode === Key.ENTER) {
            const body = {"command": command}
            const newHistory = Object.assign([], history)
            newHistory.splice(currentPos, 0, command)

            fetchApi(`/commands`, {method: 'POST', body: JSON.stringify(body)})
            .then(response => response.json())
            .then(data => {
                setHistory(newHistory)
                setCurrentPos(currentPos + 1)

                if (data.result == "") return
                setResults([...results, {"id": results.length, "result": data.result}])
            })

            setCommand("")
        }
        if (e.keyCode === Key.UP) {
            if (currentPos > 0) {
                setCommand(history[currentPos - 1])
                setCurrentPos(currentPos - 1)
            }
        }
        if (e.keyCode === Key.DOWN) {
            setCommand(history[currentPos])
            if (currentPos < history.length -1) {
                setCurrentPos(currentPos + 1)
            }
        }
        e.stopPropagation()
    }

    return (
        <>
            <div className="command">
                Command Result
            </div>
            <div className="command-result">
                {results.map(result => (
                    <div key={result.id}>
                        {result.result}
                    </div>
                ))}
            </div>
            <div className={classnames('command-input input-group')}>
                <span className="input-group-addon">
                    <i className={'fa fa-fw fa-terminal'}/>
                </span>
                <input
                    type="text"
                    placeholder="Enter command"
                    className="form-control"
                    value={command}
                    onChange={onChange}
                    onKeyDown={onKeyDown}
                />
            </div>
        </>
    )
}