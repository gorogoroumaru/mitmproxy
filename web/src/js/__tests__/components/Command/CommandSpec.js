import React from 'react'
import renderer from 'react-test-renderer'
import CommandBar from '../../../components/CommandBar'

describe('CommandBar Component', () => {
    let commandBar = renderer.create(
            <CommandBar />),
        tree = commandBar.toJSON()

    it('should render correctly', () => {
        expect(tree).toMatchSnapshot()
    })
})
