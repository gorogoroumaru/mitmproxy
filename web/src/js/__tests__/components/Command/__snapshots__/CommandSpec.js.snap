// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandBar Component should render correctly 1`] = `
Array [
  <div
    className="command"
  >
    Command Result
  </div>,
  <div
    className="command-result"
  />,
  <div
    className="command-input input-group"
  >
    <span
      className="input-group-addon"
    >
      <i
        className="fa fa-fw fa-terminal"
      />
    </span>
    <input
      className="form-control"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter command"
      type="text"
      value=""
    />
  </div>,
]
`;
